src/app/clusters/groups-server/groups-server.cpp

```diff
@@ -235,59 +264,59 @@ bool emberAfGroupsClusterViewGroupCallback(app::CommandHandler * commandObj, con
 bool emberAfGroupsClusterGetGroupMembershipCallback(app::CommandHandler * commandObj, const app::ConcreteCommandPath & commandPath,
                                                     const Commands::GetGroupMembership::DecodableType & commandData)
 {
+    auto fabricIndex  = GetFabricIndex(commandObj);
+    auto endpointId   = commandPath.mEndpointId;
     auto & groupCount = commandData.groupCount;
     auto & groupList  = commandData.groupList;
+    auto * groups     = GetGroupDataProvider();
+    auto * groupIt    = groups ? groups->IterateGroupMappings(fabricIndex, endpointId) : nullptr;
 
     EmberStatus status = EMBER_ZCL_STATUS_FAILURE;
-    uint8_t i, j;
-    uint8_t count = 0;
-    uint8_t list[EMBER_BINDING_TABLE_SIZE << 1];
+    uint8_t count      = 0;
+    uint8_t list[CHIP_CONFIG_MAX_GROUPS_PER_FABRIC << 1];
     uint8_t listLen = 0;
     CHIP_ERROR err  = CHIP_NO_ERROR;
 
     emberAfGroupsClusterPrint("RX: GetGroupMembership [");
 
+    if (nullptr == groupIt)
+    {
+        count = 0;
+    }
     // For all networks, Get Group Membership commands may be sent either
     // unicast or broadcast (removing the ZLL-specific limitation to unicast).
 
     // When Group Count is zero, respond with a list of all active groups.
     // Otherwise, respond with a list of matches.
     // TODO: https://github.com/project-chip/connectedhomeip/issues/10335
-    if (groupCount == 0)
+    else if (groupCount == 0)
     {
-        for (i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+        // When Group Count is zero, respond with a list of all active groups.
+        GroupDataProvider::GroupMapping mapping;
+        while (groupIt->Next(mapping))
         {
-            EmberBindingTableEntry entry;
-            status = emberGetBinding(i, &entry);
-            if ((status == EMBER_SUCCESS) && (entry.type == EMBER_MULTICAST_BINDING) && (entry.local == emberAfCurrentEndpoint()))
-            {
-                list[listLen]     = EMBER_LOW_BYTE(entry.groupId);
-                list[listLen + 1] = EMBER_HIGH_BYTE(entry.groupId);
-                listLen           = static_cast<uint8_t>(listLen + 2);
-                count++;
-            }
+            list[listLen]     = EMBER_LOW_BYTE(mapping.group);
+            list[listLen + 1] = EMBER_HIGH_BYTE(mapping.group);
+            listLen           = static_cast<uint8_t>(listLen + 2);
+            count++;
         }
     }
     else
     {
+        GroupDataProvider::GroupMapping mapping;
         auto iter = groupList.begin();
         while (iter.Next())
         {
             GroupId groupId = iter.GetValue();
             emberAfGroupsClusterPrint(" 0x%02" PRIx16, groupId);
-            for (j = 0; j < EMBER_BINDING_TABLE_SIZE; j++)
+            while (groupIt->Next(mapping))
```

Comment:
I don't think this behaves correctly...  It needs to restart iteration at the beginning for every new index in the incoming list, no?

We desperately need actual tests for this code.  :(

Reply:
Correct again. I re-wrote the function to avoid the EMBER_BINDING_TABLE_SIZE limit, and added YAML tests for the GetGroupMembership command.